src/mongodb_scram_sha256_fmt_plug.c

```diff
@@ -0,0 +1,303 @@
+/*
+ * This software is Copyright (c) 2025, Dhiru Kholia <dhiru.kholia at gmail.com>,
+ * and it is hereby released to the general public under the following terms:
+ * Redistribution and use in source and binary forms, with or without modification,
+ * are permitted.
+ *
+ * https://github.com/mongodb/specifications/blob/master/source/auth/auth.md
+ * https://datatracker.ietf.org/doc/html/rfc7677
+ */
+
+#if FMT_EXTERNS_H
+extern struct fmt_main fmt_mongodb_scram_sha256;
+#elif FMT_REGISTERS_H
+john_register_one(&fmt_mongodb_scram_sha256);
+#else
+
+#include <string.h>
+
+#ifdef _OPENMP
+#include <omp.h>
+#endif
+
+#include "arch.h"
+#include "misc.h"
+#include "memory.h"
+#include "common.h"
+#include "formats.h"
+#include "johnswap.h"
+#include "sha.h"
+#include "base64_convert.h"
+#include "hmac_sha.h"
+
+#undef SIMD_COEF_32 // FIXME
+#include "simd-intrinsics.h"
+#include "pbkdf2_hmac_sha256.h"
+
+#if defined SIMD_COEF_32
+#define SIMD_KEYS		(SIMD_COEF_32 * SIMD_PARA_SHA256)
+#endif
+
+#define FORMAT_LABEL            "scram-sha256"
+#define FORMAT_NAME             ""
+#define ALGORITHM_NAME          "MongoDB ServerKey SCRAM PBKDF2-SHA256 " SHA256_ALGORITHM_NAME
```

Comment:
Let's drop `MongoDB ServerKey ` part from this string. This looks like a generic SCRAM format, so probably usable not only for MongoDB. The SHA-1 kind of this format doesn't have that in its algorithm name.

A `Server Key` string is part of the SCRAM spec, no need to make it into the algorithm name string.

Oh, I see that the other format uses `Client Key`. Both are part of the spec, but to be used by client vs. server. Why is the difference here? Maybe we want to make both formats more generic by including this client vs. server bit in the hash strings?

Separately, we may want to rename these format files removing the `mongodb_` part. Or we could make them into one shared format with SHA-1 vs. SHA-256 becoming a tunable cost.

Reply:
OK, per that link MongoDB stores both `Stored Key` (for authenticating clients) and `Server Key` (to let clients authenticate server), and you choose to attack the latter because it's slightly quicker. Yet some other service may use SCRAM for client authentication only and wouldn't have/store `Server Key`. So we could support both modes maybe?

Either way (supporting both or not), we could do it in the same way for both formats (SHA-1 and SHA-256), right?

Reply:
A separate later project could be to add support for attacking SCRAM challenge/response messages - also either/both kinds (`ClientProof` or `ServerSignature`). We may want to keep this future addition in mind when we name things, but I don't know how this affects the naming. One way could be to include the names `ClientKey`, `ServerKey`, `ClientProof`, or `ServerSignature` as a field in the hash strings, or we can use numeric values for them (0 to 3). For challenge/responses, we'd also need an extra field `Auth`. https://en.wikipedia.org/wiki/Salted_Challenge_Response_Authentication_Mechanism#Proofs

Reply:
Oh, I just realized that the older format is MongoDB-specific - it includes preprocessing with MD5 and a ":mongo:" string. So we can't just make them consistent. Since it's unlikely that new uses of SCRAM with SHA-1 will appear now, maybe we can keep the older format MongoDB-specific, but maybe we should rename its label and format tag to reflect that then. You could also want to switch it to using `ServerKey` if that's available in those older versions of MongoDB too?